### PR TITLE
add i18n tests

### PR DESCRIPTION
**Pull request by mattbjordan**
Adding a bunch of tests for the i18n package

Some questions and notes:
- The only functionality change is that `getEnglishString` will consider both `en` _and_ `en-US` locales instead of _only_ `en`.
  - Side note: This is the default functionality for i18next too.
- There is currently a skipped test related to `getEnglishString` that I'd like feedback about.
  - **I want to make a breaking change** (but have not yet).
  - This `getEnglishString` `namespace` parameter doesn't make sense. I think there should be `getEnglishString` and `getEnglishStringWithNamespace` methods to mirror how the `getLocalizedString` and `getLocalizedStringWithNamespace` methods work. Basically, the WithNamespace methods will use the passed in namespace regardless of any i18next options passed in, while the "normal" method allows the options to work normally. 
- Does anyone know if the namespace promises are supposed to do anything other than make you `await` them?
  - I'm not sure how to test the `getNamespacePromise` and `unregisterNamespace` methods as the promises are just empty or null upon returning/resolving.
  - And `unregisterNamespace` doesn't seem to actually do anything relevant anyway - and isn't used. It doesn't effect the ability to use a namespace. i18next has no way to unload a namespace.
  
A separate note: I think the changes to `common/changes/@itwin/geonames-extension/node-types-18_2022-09-14-15-18.json` came from a weird merge, but I couldn't get `rush change` to let me remove them.

### CODE REVIEW

File: common/changes/@itwin/geonames-extension/node-types-18_2022-09-14-15-18.json [hunk 1]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (aruniverse) [inline common/changes/@itwin/geonames-extension/node-types-18_2022-09-14-15-18.json:1]: merge with master, then you should be able to nuke this file

Comment (mattbjordan) [reply common/changes/@itwin/geonames-extension/node-types-18_2022-09-14-15-18.json:1]: It's already merged and up-to-date with master, which is why I am confused by it. This change file doesn't exist in master. Perhaps the changes were reverted but the change file still stayed in my branch?

File: core/i18n/src/ITwinLocalization.ts [hunk 2]
```diff
@@ -176,7 +172,7 @@ export class ITwinLocalization implements Localization {
       throw new Error("Translation key must map to a string, but the given options will result in an object");
     }
 
-    const en = this.i18next.getFixedT("en", namespace);
+    const en = this.i18next.getFixedT(["en", "en-US"], namespace);
```
Comment (aruniverse) [inline core/i18n/src/ITwinLocalization.ts:175]: whats the reason for this change?

as i understand, the `en-US` gets added due to `detectOptions.order: ["querystring", "navigator", "htmlTag"]`
`en` is the default fallback we set, and all our locale files are for `en`

Comment (mattbjordan) [reply core/i18n/src/ITwinLocalization.ts:175]: I just thought it made more sense to also check `en-US` as a backup when specifically trying to `getEnglishString()`. 

`getFixedT("en", ...)` will only check for the localization files in `locales/en/`. `getFixedT(["en", "en-US"], ...)` will check in `locales/en/` first, then `locales/en-US` as a backup. If there's a reason we only want the former, then lmk and I'll revert it.

But yes, typically for `getLocalizedString`, the language is detected as `en-US` and we set `en` as the backup.

Comment (aruniverse) [reply core/i18n/src/ITwinLocalization.ts:175]: tbh i never understood why we as a library gave preference to english, and `getEnglishString` was apart of the Localization interface.

I assumed its because we needed english strings for the keyins in core-frontend, but even then, the fn shouldve been optional in the interface imo

if we're adding en-us, why arent we adding the [rest](https://github.com/ladjs/i18n-locales/blob/master/index.json#L74-L87)? i just dont see a need for this

maybe @calebmshafer or @evelynpreslar-bentley can chime in

File: common/changes/@itwin/geonames-extension/node-types-18_2022-09-14-15-18.json (hunk 1 above)
Comment (aruniverse) [reply common/changes/@itwin/geonames-extension/node-types-18_2022-09-14-15-18.json:1]: try deleting the file, the pkg doesnt exist anymore, so you should be able to pass `rush change -v` w/o this

Comment (mattbjordan) [reply common/changes/@itwin/geonames-extension/node-types-18_2022-09-14-15-18.json:1]: Everything I've tried with `rush change` still doesn't pass. The error is that it can't find the change file that I am trying to delete..

Comment (pmconne) [reply common/changes/@itwin/geonames-extension/node-types-18_2022-09-14-15-18.json]: I deleted it and pushed.

File: core/i18n/src/ITwinLocalization.ts (hunk 2 above)
Comment (calebmshafer) [reply core/i18n/src/ITwinLocalization.ts:175]: Let's leave it as is for now.

The reason for english preference is due to key-ins and before we add en-US I'd want to consider if there's something else we should be doing instead.

**Comment by pmconne** [comment]
Will this have any effect on the following error I always observe when running the tests? On linux. Usu I use `build:ci` to build them.
```
--[ FAILURE: @itwin/core-i18n ]------------------------------[ 1.00 seconds ]--

/home/paul/c/j2/s/itwinjs/tools/certa/lib/CertaConfig.js:71
            throw new Error(`The specified testBundle file "${resolvedOpts.testBundle}" does not exist.`);
            ^

Error: The specified testBundle file "/home/paul/c/j2/s/itwinjs/core/i18n/lib/cjs/test/webpack/bundled-tests.js" does not exist.
    at fromObject (/home/paul/c/j2/s/itwinjs/tools/certa/lib/CertaConfig.js:71:19)
    at Object.fromConfigFile (/home/paul/c/j2/s/itwinjs/tools/certa/lib/CertaConfig.js:88:16)
    at Object.<anonymous> (/home/paul/c/j2/s/itwinjs/tools/certa/bin/certa.js:87:24)
    at Module._compile (node:internal/modules/cjs/loader:1126:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1180:10)
    at Module.load (node:internal/modules/cjs/loader:1004:32)
    at Function.Module._load (node:internal/modules/cjs/loader:839:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12)
    at node:internal/main/run_main_module:17:47
```

File: core/i18n/src/test/ITwinLocalization.test.ts [hunk 3]
```diff
@@ -146,7 +508,583 @@ describe("ITwinLocalization", () => {
...
+       * This functionality doesn't make sense. It should work the opposite like how getLocalizedStringWithNamespace() does.
+       * getEnglishString() shouldn't take in a namespace. The 'namespace' parameter should be removed.
+       * A getEnglishStringWithNamespace() method could be created.
+       * This is a breaking change, so is left as a to-do until discussed and approved.
+       */
+      it.skip("given namespace DOES NOT override namespace translation option", () => {
+        assert.equal(localization.getEnglishString("Default", "FirstTrivial", { ns: "Test" }), "First level string (test)");
+        assert.equal(localization.getEnglishString("Default", "FirstSubstitution1", { str: "CUSTOM1", ns: "Test" }), "First level CUSTOM1 (test)");
+      });
+    });
```
Comment (mattbjordan) [inline core/i18n/src/test/ITwinLocalization.test.ts:940]: @calebmshafer or @wgoehrig What do you think about my comment here? `getEnglishString()` takes in a `namespace` parameter which doesn't really make sense. I'd think it should be like `getLocalizedString()` and `getLocalizedStringWithNamespace()` where there is another function that handles providing a namespace parameter.

For `getLocalizedStringWithNamespace()`, the passed in namespace is used no matter what i18next options are passed in (see this [test](https://github.com/iTwin/itwinjs-core/blob/552c2fa584673f170f8e2a7dbca89d49b36a1f34/core/i18n/src/test/ITwinLocalization.test.ts#L747)). For `getEnglishString()`, this is not the case.

I understand if we don't want to make these changes right now, but should I leave this test as skipped, comment it out, or just delete it?

Comment (calebmshafer) [reply core/i18n/src/test/ITwinLocalization.test.ts:940]: I think we should iron out the differences between the english and non-english methods however that cannot be done through a breaking change.

Adding a new method would be fine, is there anywhere it will actually be used in the code?

File: core/i18n/src/ITwinLocalization.ts [hunk 4]
```diff
@@ -143,28 +143,24 @@ export class ITwinLocalization implements Localization {
...
-    return this.getLocalizedString(fullKey, options);
+    options = {
+      ...options,
+      ns: namespace,
+    };
+    return this.getLocalizedString(key, options);
   }
 
   /** Gets the English translation.
+   * @note Checks locales 'en', 'en-US' - in that order
```
Comment (aruniverse) [inline core/i18n/src/ITwinLocalization.ts:163]: revert this line too

Comment (mattbjordan) [reply core/i18n/src/ITwinLocalization.ts]: Good catch. Done.

**Comment by mattbjordan** [comment]
> Will this have any effect on the following error I always observe when running the tests locally? On linux. Usu I use `build:ci` to build them.

@pmconne I wouldn't expect these changes to affect that error. However, I believe you just need to run `webpack:test` after building and before testing to fix that.

**Comment by pmconne** [comment]
> you just need to run `webpack:test` after building and before testing

Please make sure either the `build` or the `test` script does the webpacking. `rush build:ci` followed by `rush cover` should suffice to run all tests.

**Comment by aruniverse** [comment]
> Please make sure either the `build` or the `test` script does the webpacking. `rush build:ci` followed by `rush cover` should suffice to run all tests.

That is not part of the scope of this pr, and should be done in a separate one if we choose to do so. I had broken this out of the build:ci command a few months ago to save time for all devs & pipelines where we don't need to waste time webpacking the tests.

**Comment by pmconne** [comment]
> should be done in a separate one if we choose to do so

Separate PR fine but don't make every dev run webpack:tests for this one package when they want to run all the tests.

**Comment by mattbjordan** [comment]
> Please make sure either the build or the test script does the webpacking. rush build:ci followed by rush cover should suffice to run all tests

Should I add webpack:test to `build`, `build:ci`, or both? (in a separate PR)

**Comment by pmconne** [comment]
> Should I add webpack:test to build, build:ci, or both? 

build:ci.

File: core/i18n/src/test/ITwinLocalization.test.ts (hunk 3 above)
Comment (calebmshafer) [reply core/i18n/src/test/ITwinLocalization.test.ts]: Yea, even though it's not the most ideal the most important part is that we don't break the API as it is today.

File: core/i18n/src/ITwinLocalization.ts [hunk 5]
```diff
@@ -143,25 +143,20 @@ export class ITwinLocalization implements Localization {
     return value;
   }
 
-  /** Similar to `getLocalizedString` but the namespace is a separate param and the key does not include the namespace.
+  /** Similar to `getLocalizedString` but the default namespace is a separate parameter and the key does not need
```
Comment (calebmshafer) [inline core/i18n/src/ITwinLocalization.ts:146]: Instead of changing the functionality of this method. Let's just remove it. It's marked as internal and we should clean this up since no one should be using it.

FYI @raplemie  @GerardasB this is used in AppUI so we'll be updating it to use the correct method. The same changes will need to be made for 4.0.

Comment (raplemie) [reply core/i18n/src/ITwinLocalization.ts:146]: Added https://github.com/iTwin/itwinjs-backlog/issues/510 for tracking

Comment (mattbjordan) [reply core/i18n/src/ITwinLocalization.ts:146]: I've pushed a commit that reverts changes to `getLocalizedStringWithNamespace` and skips a few edge-case tests that fail without the better code. This will preserve the method to make sure that it will work (because no changes were made) up until it is removed. The removal of the method will be in a separate PR as it is a bigger and breaking change - leaving this PR focused on tests - and is tracked by [this issue](https://github.com/iTwin/itwinjs-backlog/issues/513).

**Review by calebmshafer — APPROVED** [review]
